### PR TITLE
Pythonize/open ems

### PR DESCRIPTION
**Pull request by SengerM**
Interface is more robust and user friendly. Did it only for some functions I am using.

## Important
In `openEMS.SetBoundaryCond` I changed the interface, instead of receiving a list with 6 elements, it receives the 6 elements as function arguments. Advantages:
- It is obvious from the source code what is each element.
- Makes it possible to use named arguments with any arbitrary ordering.
- More robust.
The conversion from any code running with the previous interface to the new interface should be straightforward:
```python
ems.SetBoundaryCond(my_BC)
```
becomes
```python
ems.SetBoundaryCond(*my_BC)
```
which should be easy to do with a single find-and-replace `.SetBoundaryCond(`→`.SetBoundaryCond(*`.

All the rest I think should be backwards compatible.

### CODE REVIEW

File: python/openEMS/openEMS.pyx [hunk 1]
```diff
@@ -458,26 +465,29 @@ cdef class openEMS:
                     continue
                 grid.AddLine(n, hint[n])
 
-    def Run(self, sim_path, cleanup=False,setup_only=False, debug_material=False, debug_pec=False,
-            debug_operator=False, debug_boxes=False, debug_csx=False, verbose=None, **kw):
-        """ Run(sim_path, cleanup=False, setup_only=False, verbose=None)
-
-        Run the openEMS FDTD simulation.
+    def Run(self, sim_path:str|Path, cleanup:bool=False, setup_only:bool=False, debug_material:bool=False, debug_pec:bool=False, debug_operator:bool=False, debug_boxes:bool=False, debug_csx:bool=False, verbose:int=0, numThreads:int=0):
```
Comment (thliebig) [inline python/openEMS/openEMS.pyx:468]: We will need the **kw argument as a recent PR added the "_SetLibraryArguments" method...

File: python/openEMS/openEMS.pyx [hunk 2]
```diff
@@ -271,22 +277,26 @@ cdef class openEMS:
         """
         self.thisptr.SetCustomExcite(_str, f0, fmax)
 
-    def SetBoundaryCond(self, BC):
-        """ SetBoundaryCond(BC)
+    def SetBoundaryCond(self, x_min:str|int, x_max:str|int, y_min:str|int, y_max:str|int, z_min:str|int, z_max:str|int):
```
[the remark this answers is not in the document]
Comment (SengerM) [reply python/openEMS/openEMS.pyx]: Sounds good. See my new commits.

File: python/openEMS/openEMS.pyx [hunk 3]
```diff
@@ -112,19 +118,19 @@ cdef class openEMS:
...
+            raise ValueError(f'`time_steps` must be > 0, received {time_steps}. ')
+        self.thisptr.SetNumberOfTimeSteps(int(time_steps))
 
-        Set the end criteria value. E.g. 1e-6 for -60dB
-        """
-        self.thisptr.SetEndCriteria(val)
+    def SetEndCriteria(self, energy_decay_factor:float):
+        """Set the end criterion value. For example, `energy_decay_factor=1e-5` will make the simulation stop if energy has decayed by a factor of `1e5`. `1e-6` for -60dB."""
+        if not isinstance(energy_decay_factor, float):
+            raise TypeError(f'`energy_decay_factor` must be an instance of `float`, received object of type {type(energy_decay_factor)}. ')
```
[the remark this answers is not in the document]
Comment (SengerM) [reply python/openEMS/openEMS.pyx:135]: Could be, it is just that in the direct interface with the user I usually like to provide very explicit error messages, thus making it trivial for a user to fix issues. But letting `float(energy_decay_factor)` handle this should also work and is probably explicit enough.

File: python/openEMS/openEMS.pyx (hunk 1 above)
Comment (SengerM) [reply python/openEMS/openEMS.pyx:500]: I would fix that in the merge process between this and that. 

# About the need of `**kw`
If there is a bunch of arguments that will be specifically given to some method, like `_SetLibraryArguments`, I would implement that with its own dictionary, like this (without any `**kw`):
```python
def Run(self, sim_path, ..., numThreads:int=0, arguments_given_to_SetLibraryArguments:dict=None):
   ...
   if arguments_given_to_SetLibraryArguments is not None:
      self._SetLibraryArguments(**arguments_given_to_SetLibraryArguments)
```
and in the user side it looks like
```python
simulation.Run(
    sim_path = path_to_wherever/'simulation',
    ...
    arguments_given_to_SetLibraryArguments = dict(
        argument_1 = 'whatever',
        argument_2 = 'hi!',
        ...
    ),
)
```
#### Some advantages I see:
- The arguments for `_SetLibraryArguments` remain grouped at all times.
- The validation of that whole dictionary becomes a sole responsibility of `_SetLibraryArguments`.
- In `Run`, the documentation of `arguments_given_to_SetLibraryArguments` is literally "See `_SetLibraryArguments`.".
- Implementation is trivial, as shown above. 
- Maintenance is also trivial. Any change in `_SetLibraryArguments` has literally no effect in the implementation of `Run`.
- No one needs to use the ambiguous `**kw`, neither `Run` nor `_SetLibraryArguments`. Both functions have clear signatures.
- For the user it is obvious which set of arguments is passed to `_SetLibraryArguments`. Should `_SetLibraryArguments` show up in the error stack trace, he/she knows that the issue is inside `arguments_given_to_SetLibraryArguments`.

#### Disadvantages I see:
- The user has to type one more line of code.

[Plotly](https://plotly.com/python/) has this kind of approach, I think it works well, I like it.

Comment (thliebig) [reply python/openEMS/openEMS.pyx:500]: Well that would certainly work, but I prefer it simple, and 

`FDTD.Run(Sim_Path, engine='basic'`

just reads much nicer, flows more with the other options, the user does not need to know where every options goes and why and is much cleaner...

Please restore the version with **kw if you want this PR merged, thanks

**Comment by thliebig** [comment]
The fast majority of this changes I have simply to decline. Thus I close this.

**Pull request closed without merging**